Use different keychain attribute for passkey credential ID

aznair opened the pull request:
#### 1dc9ae6c18649e69843343042de19d3042e64d2e
<pre>
Use different keychain attribute for passkey credential ID
<a href="https://bugs.webkit.org/show_bug.cgi?id=281344">https://bugs.webkit.org/show_bug.cgi?id=281344</a>
<a href="https://rdar.apple.com/137771569">rdar://137771569</a>

Reviewed by Brent Fulgham.

Use kSecAttrAlias going forward, and if a query using this field fails then fall
back to kSecAttrApplicationLabel

* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
(getAllLocalAuthenticatorCredentialsImpl):
(+[_WKWebAuthenticationPanel deleteLocalAuthenticatorCredentialWithGroupAndID:credential:]):
(+[_WKWebAuthenticationPanel setDisplayNameForLocalCredentialWithGroupAndID:credential:displayName:]):
(+[_WKWebAuthenticationPanel setNameForLocalCredentialWithGroupAndID:credential:name:]):
(+[_WKWebAuthenticationPanel importLocalAuthenticatorWithAccessGroup:credential:error:]):
(+[_WKWebAuthenticationPanel exportLocalAuthenticatorCredentialWithGroupAndID:credential:error:]):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/LocalAuthenticator.mm:
(WebKit::LocalAuthenticatorInternal::produceHashSet):
(WebKit::LocalAuthenticator::getExistingCredentials):
(WebKit::LocalAuthenticator::processLargeBlobExtension):
(WebKit::LocalAuthenticator::continueMakeCredentialAfterUserVerification):
(WebKit::LocalAuthenticator::continueGetAssertionAfterUserVerification):
(WebKit::LocalAuthenticator::receiveException const):
(WebKit::LocalAuthenticator::deleteDuplicateCredential const):
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::cleanUpKeychain):
(WTR::TestController::keyExistsInKeychain):

Canonical link: <a href="https://commits.webkit.org/285575@main">https://commits.webkit.org/285575@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/afd314606349cf5522c1aa9e0f656d88376999b6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73078 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52507 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25886 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77293 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24316 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75193 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/60312 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/289 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57434 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15916 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76145 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47448 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62893 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37851 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44087 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20363 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22645 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/66213 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65943 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20718 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78957 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/72335 "Built successfully and passed tests") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65870 "Passed tests") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/438 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62895 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65147 "Passed tests") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7143 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/94114 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11258 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/343 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3094 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20700 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/375 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/371 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/370 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/LocalAuthenticator.mm:
```diff
@@ -359,18 +361,24 @@ static inline uint8_t authDataFlags(ClientDataType type, LocalConnection::UserVe
             (id)kSecReturnAttributes: @YES,
             (id)kSecReturnPersistentRef: @YES,
-        };
+        } mutableCopy];
```
This is a memory leak, since `mutableCopy` returns a +1 object. Please use `RetainPtr` and `adoptNS`.

aznair replied:
I've used the `adoptNS` - `setDictionary` pattern I saw elsewhere in this code.

pxlcoder commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/LocalAuthenticator.mm:
```diff
@@ -730,18 +749,24 @@ static inline uint8_t authDataFlags(ClientDataType type, LocalConnection::UserVe
+        (id)kSecAttrAlias: nsCredentialId.get(),
         (id)kSecUseDataProtectionKeychain: @YES
-    };
+    } mutableCopy];
```
Ditto re: memory leak.

aznair replied:
Done

pxlcoder commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/LocalAuthenticator.mm:
```diff
@@ -792,14 +823,20 @@ static inline uint8_t authDataFlags(ClientDataType type, LocalConnection::UserVe
             (id)kSecAttrSynchronizable: (id)kSecAttrSynchronizableAny,
             (id)kSecUseDataProtectionKeychain: @YES
-        };
+        } mutableCopy];
```
Ditto re: memory leak.

aznair replied:
Done

pxlcoder commented on Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
```diff
@@ -608,22 +608,36 @@ void initializeWebViewConfiguration(const char* libraryPath, WKStringRef injecte
+        (id)kSecAttrAlias: credentialID.get(),
         (id)kSecUseDataProtectionKeychain: @YES
-    };
+    } mutableCopy];
```
One more place 😅

aznair replied:
Done

brentfulgham commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -256,6 +265,10 @@ - (void)setDelegate:(id<_WKWebAuthenticationPanelDelegate>)delegate
+    BOOL useAlternateKeychainAttribute = NO;
+#if USE(APPLE_INTERNAL_SDK)
+    useAlternateKeychainAttribute = shouldUseAlternateKeychainAttribute();
+#endif
```
You repeat this block of code a half dozen times. I think it would be cleaner if you defined `shouldUseAlternateKeychainAttribute()` to do this work so you always just called that one function:

```
static BOOL shouldUseAlternateKeychainAttribute()
{
#if USE(APPLE_INTERNAL_SDK)
    if (![WebKit::getASCWebKitSPISupportClass() respondsToSelector:@selector(shouldUseAlternateKeychainAttribute)])
        return NO;

    return [WebKit::getASCWebKitSPISupportClass() shouldUseAlternateKeychainAttribute];
#else
    return NO;
#endif
}
```

aznair replied:
Done

brentfulgham commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -268,13 +281,24 @@ - (void)setDelegate:(id<_WKWebAuthenticationPanelDelegate>)delegate
+        if (useAlternateKeychainAttribute)
+            [query setObject:byCredentialId forKey:bridge_cast(kSecAttrAlias)];
+        else
+            [query setObject:byCredentialId forKey:bridge_cast(kSecAttrApplicationLabel)];
```
You repeat this block of code a bunch of times, too. Why not use something like:

```
CFStringRef credentialKey = useAlternateKeychainAttribute ? kSecAttrAlias : kSecAttrApplicationLabel;
[query setObject:byCredentialId forKey:bridge_cast(credentialKey)];
```

aznair replied:
I've applied this to all the places where I add one key or the other to a dictionary

brentfulgham commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -384,16 +411,30 @@ + (void)deleteLocalAuthenticatorCredentialWithID:(NSData *)credentialID
+    if (useAlternateKeychainAttribute) {
+        [deleteQuery removeObjectForKey:bridge_cast(kSecAttrAlias)];
+        [deleteQuery setObject:credentialID forKey:bridge_cast(kSecAttrApplicationLabel)];
+        SecItemDelete((__bridge CFDictionaryRef)deleteQuery.get());
```
Should you log the status if/when this fails?

aznair replied:
We haven't logged delete failures so far so I didn't want to deviate from that.

brentfulgham commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/LocalAuthenticator.mm:
```diff
@@ -60,6 +60,14 @@
+    if (![WebKit::getASCWebKitSPISupportClass() respondsToSelector:@selector(shouldUseAlternateKeychainAttribute)])
+        return NO;
+
+    return [WebKit::getASCWebKitSPISupportClass() shouldUseAlternateKeychainAttribute];
```
Why are you duplicating all of this code from `Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm`? You should have one copy of this code in a place you can use from both of these compilation units.

aznair replied:
I put it in LocalAuthenticator.h within the WebKit namespace.

brentfulgham commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -384,16 +411,30 @@ + (void)deleteLocalAuthenticatorCredentialWithID:(NSData *)credentialID
+    if (useAlternateKeychainAttribute)
+        [deleteQuery setObject:credentialID forKey:(__bridge id)kSecAttrAlias];
+    else
+        [deleteQuery setObject:credentialID forKey:(__bridge id)kSecAttrApplicationLabel];
```
You should assign `kSecAttrAlias` or `kSecAttrApplicationLabel` to a temporary based on a conditional, and avoid all of this bulky code duplication.

aznair replied:
Done

brentfulgham commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -685,12 +772,24 @@ + (NSData *)exportLocalAuthenticatorCredentialWithGroupAndID:(NSString *)group c
+    if (useAlternateKeychainAttribute)
+        [query setObject:credentialID forKey:(id)kSecAttrAlias];
+    else
+        [query setObject:credentialID forKey:(id)kSecAttrApplicationLabel];
```
Ditto my earlier comments about all of this duplicated code.

aznair replied:
Done

brentfulgham reviewed (approved):
I think this patch needs some cleanup, but the overall approach is good. Please tidy this up before landing.

pxlcoder commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/LocalAuthenticator.mm:
```diff
@@ -86,10 +86,20 @@ static bool shouldUpdateQuery()
 
+BOOL shouldUseAlternateKeychainAttribute()
+{
+#if USE(APPLE_INTERNAL_SDK)
```
`USE(APPLE_INTERNAL_SDK)` in an implementation file is an anti-pattern.

Since you already have the `respondsToSelector` check, why is it necessary? The method could be forward declared for Open Source builds.

brentfulgham replied:
Aditya has a good point. We should probably be gating this with a `HAVE()` macro, though I think we can just forward declare. Other uses of `getASCWebKitSPISupportClass()` happen outside of `USE(APPLE_INTERNAL_SDK)` in WebAuthn code, so using the compile guard seems wrong.

brentfulgham replied:
And add `shouldUseAlternateKeychainAttribute` to the forward declarations in `WebKit/platform/spi/Cococa/AuthenticationServiesCoreSPI.h`

aznair replied:
The forward declaration of `ASCWebKitSPISupport` is gated behind an `||` of three feature flags. I've gone with `HAVE(UNIFIED_ASC_AUTH_UI)` for my code here to replace `USE(APPLE_INTERNAL_SDK)`

brentfulgham commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/LocalAuthenticator.mm:
```diff
@@ -503,7 +531,6 @@ static inline uint8_t authDataFlags(ClientDataType type, LocalConnection::UserVe
         credentialId = digest->computeHash();
         m_provisionalCredentialId = toNSData(credentialId);
 
-#if ASSERT_ENABLED
```
Yikes!

jasl23 committed:
Use different keychain attribute for passkey credential ID
https://bugs.webkit.org/show_bug.cgi?id=281344
rdar://137771569

Reviewed by Brent Fulgham.

Use kSecAttrAlias going forward, and if a query using this field fails then fall
back to kSecAttrApplicationLabel

* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
(getAllLocalAuthenticatorCredentialsImpl):
(+[_WKWebAuthenticationPanel deleteLocalAuthenticatorCredentialWithGroupAndID:credential:]):
(+[_WKWebAuthenticationPanel setDisplayNameForLocalCredentialWithGroupAndID:credential:displayName:]):
(+[_WKWebAuthenticationPanel setNameForLocalCredentialWithGroupAndID:credential:name:]):
(+[_WKWebAuthenticationPanel importLocalAuthenticatorWithAccessGroup:credential:error:]):
(+[_WKWebAuthenticationPanel exportLocalAuthenticatorCredentialWithGroupAndID:credential:error:]):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/LocalAuthenticator.mm:
(WebKit::LocalAuthenticatorInternal::produceHashSet):
(WebKit::LocalAuthenticator::getExistingCredentials):
(WebKit::LocalAuthenticator::processLargeBlobExtension):
(WebKit::LocalAuthenticator::continueMakeCredentialAfterUserVerification):
(WebKit::LocalAuthenticator::continueGetAssertionAfterUserVerification):
(WebKit::LocalAuthenticator::receiveException const):
(WebKit::LocalAuthenticator::deleteDuplicateCredential const):
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::cleanUpKeychain):
(WTR::TestController::keyExistsInKeychain):

Canonical link: https://commits.webkit.org/285575@main
